### PR TITLE
Authentication enabled

### PR DESCRIPTION
**Pull request by jamesc**
## Purpose

* Adds authentication to the sample.  Gives a simple example of using EasyAuth authentication with either workforce identity (Entra ID) or customer identity (Entra External ID).

It extends the pre-/post- hook pattern from https://github.com/Azure-Samples/azure-search-openai-demo/pull/891 to also work with External ID via an additional script that is run before the `azd up` to set up an application in the External ID tenant that can create all the additional configuration (userflows, admin consent grant) that is needed in a External ID application.

## Does this introduce a breaking change?

```
[ ] Yes
[X ] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Follow instructions in README for enabling authentication with either Microsoft Entra ID or Microsoft Entra External ID

### CODE REVIEW

**Comment by pamelafox** [comment]
Thanks James! Now that I've updated this repo to OpenAI 1.0, I can look at this PR. Do you think there's any chance of modularizing this code so that there isn't so much in this repo? Or will that end up making it too difficult for developers to modify?

File: infra/core/host/container-auth.bicep [hunk 1]
```diff
@@ -0,0 +1,35 @@
+param name string
```
Comment (pamelafox) [inline infra/core/host/container-auth.bicep:1]: Have you also added this to the azure-dev/ repo?

File: src/quartapp/chat.py [hunk 2]
```diff
@@ -25,9 +26,43 @@ async def configure_openai():
...
+# Extract the best username for display from the base64 encoded header
+# X-MS-CLIENT-PRINCIPAL. We try to find various claims in the header:
+# - name
+# - preferred_username
+#
+# Fallback to X-MS-CLIENT-PRINCIPAL-NAME if none of the above are found.
+#
+# https://learn.microsoft.com/en-us/azure/app-service/configure-authentication-user-identities#decoding-the-client-principal-header
+#
+def extract_username(headers, default_username="You"):
```
Comment (pamelafox) [inline src/quartapp/chat.py:38]: Would this code work for authentication on any Azure host? (I'm curious how generically useful it is.. should it be in msal-extensions?)

File: infra/core/host/container-auth.bicep (hunk 1 above)
Comment (jamesc) [reply infra/core/host/container-auth.bicep:1]: Just created the PR - https://github.com/Azure/azure-dev/pull/3214

**Comment by mirojs** [comment]
@jamesc @pamelafox This works for Entra ID for customers (Entra External ID for customers), right? Any plan of merging to the main? Thanks.

**Comment by pamelafox** [comment]
I have taken these changes, updated them to use the Graph SDK where possible, and added KeyVault support. I wasn't able to use the Graph Beta SDK for the new stuff (flows) due to some issues that I reported in their repo, so I left them os HTTP calls.

You can see my changes in this branch comparison view:
https://github.com/jamesc/chatgpt-quickstart/compare/auth-enabled...Azure-Samples:openai-chat-app-quickstart:pamelauth

I have put all the changes in a new repo:
https://github.com/Azure-Samples/openai-chat-app-entra-auth-builtin

After a chat with Arun, we've decided to have three separate repos to help developers find what they need:

1. https://github.com/Azure-Samples/openai-chat-app-quickstart/ for developers who just need an example of managed identity
2. https://github.com/Azure-Samples/openai-chat-app-entra-auth-builtin for developers who want to setup Auth with the built-in Auth functionlity
3. https://github.com/Azure-Samples/openai-chat-app-entra-auth-local for developers who need to be able to develop with Auth locally or a non-Azure platform, using MSAL SDK via the identity package

I will close this PR for now. Please let me know if you see any issue with my changes in the new repo and file any issues or new PRs there. The Entra External ID is quite neat, and I'll try to bring that into openai-chat-app-entra-auth-local as well when I have a chance. Thank you @jamesc !

**Pull request closed without merging**